### ISSUE TITLE
Deploy new workspaces to OPS

### ISSUE THREAD

**Issue by gchang** [MAAP-Project/ZenHub#303]
A couple of things that need to happen concurrently for the new workspaces to work in Ops:

- [x] @bsatoriu Merge the 7.25.2-maap-efs changes into the main maap branch and officially build the image. Update and deploy the `maap-k8sconfig-patch.yaml` with the official builds instead of the one on dev.
- [x] @mayadebellis Merge the changes in `entrypoint.sh` so that its able to determine the url prefix without the `MACHINE_NAME` environment variable from the kubernetes config.

I've attached a devfile that does everything I demonstrated:
* Uses my flaky s3 sidecar to mount the private/public/shared buckets (Dockerfile for that image in che-k8s ~/s3fs-sidecar-dockerfile).
* Mounts /projects/.jupyter as an ephemeral folder so that new instances of Jupyter don't open up the same tabs.
* Does *not* have the MACHINE_NAME environment variable.
* Uses a nodeport to expose the SSH service to the public.

Regarding the nodeport service, my example with the ssh-enabler workspace doesn't work because that's still on a per-pod basis so the nodeport I activate in the ssh-enabler workspace doesn't apply to other pods. The issue I had with putting a nodeport inside a workspace definition is that the second one would fail because the nodeport name was the same, basically a name collision. Kubernetes does have a `generateName` configuration that will generate a random name, but Che needs a real name. So the fix was to modify Che to emulate the `generateName` behavior and populate the name with the generateName+randomsuffix. I've also included the code to determine the correct nodeport from inside the container. I also tried to see if I could create the nodeport from inside the container by doing a POST to the kubernetes API, but I can't. I get a forbidden access error.

So some more things:

- [x] Test out the devfile and then add it to the devfile registry, replacing all the other ones that use the separate yaml file.
- [x] Update the ssh_info plugin with the new port detection code.

The nodeport detection reverts back to contacting an external server to get the machine's IP address. Because we're using ELB, we need to tell AWS every port we're opening to go through the ELB. Since that's a lot of work to open/close these ports, it's just easier to connect directly to the EC2 instance.

[devfile.yaml.zip](https://github.com/MAAP-Project/ZenHub/files/6234901/devfile.yaml.zip)
[getsshport.py.zip](https://github.com/MAAP-Project/ZenHub/files/6234902/getsshport.py.zip)

**Comment by bsatoriu** [comment]
@gchang can you please also attach the code for `registry.nasa.maap.xyz/root/jupyter_image/gchang-s3fs-sidecar:latest`? I took a screenshot during our call but it would be helpful to have the most recent version if possible. I'd like to reference it for #271

**Comment by gchang** [comment]
@bsatoriu The info was in the description above, but somewhat hidden (`Dockerfile for that image in che-k8s ~/s3fs-sidecar-dockerfile`). Anyhow, here it is:

Dockerfile
```
# Adapted from https://github.com/efrecon/docker-s3fs-client/blob/master/Dockerfile

# pinning to current version
FROM alpine:3.13.2 AS build

ARG S3FS_VERSION=v1.89

RUN apk --no-cache add \
ca-certificates \
build-base \
git \
alpine-sdk \
libcurl \
automake \
autoconf \
libxml2-dev \
libressl-dev \
fuse-dev \
curl-dev && \
git clone https://github.com/s3fs-fuse/s3fs-fuse.git && \
cd s3fs-fuse && \
git checkout tags/${S3FS_VERSION} && \
./autogen.sh && \
./configure --prefix=/usr && \
make -j && \
make install

FROM alpine

COPY --from=build /usr/bin/s3fs /usr/bin/s3fs

RUN mkdir -p /projects && \
apk --no-cache add \
  ca-certificates \
  fuse \
  libxml2 \
  libcurl \
  libgcc \
  libstdc++ \
  tini && \
s3fs --version

WORKDIR /projects

COPY *.sh /usr/local/bin/

# The default is to perform all system-level mounting as part of the entrypoint
# to then have a command that will keep listing the files under the main share.
# Listing the files will keep the share active and avoid that the remote server
# closes the connection.
ENTRYPOINT [ "tini", "-g", "--", "entrypoint.sh" ]
CMD [ "empty.sh" ]
```

entrypoint.sh
```
#!/bin/sh

# Create user directories
mkdir /s3root
s3fs maap-$MAAP_ENVIRONMENT-workspace:/ /s3root -o use_cache=/tmp -o imdsv1only -o iam_role=auto
mkdir -p /s3root/$CHE_WORKSPACE_NAMESPACE
mkdir -p /s3root/shared/$CHE_WORKSPACE_NAMESPACE
umount /s3root
rm -rf /s3root

s3fs maap-$MAAP_ENVIRONMENT-workspace:/$CHE_WORKSPACE_NAMESPACE /my-private-bucket -o use_cache=/tmp -o imdsv1only -o iam_role=auto
s3fs maap-$MAAP_ENVIRONMENT-workspace:/shared /shared-buckets -o ro -o use_cache=/tmp -o imdsv1only -o iam_role=auto
s3fs maap-$MAAP_ENVIRONMENT-workspace:/shared/$CHE_WORKSPACE_NAMESPACE /my-public-bucket -o use_cache=/tmp -o imdsv1only -o iam_role=auto
```

**Comment by bsatoriu** [comment]
@gchang I've deployed the merged `7.25.2-maap` image to ops and updated https://github.com/MAAP-Project/maap-eclipseche-ops/blob/master/che7/maap-k8sconfig-patch.yaml

**Comment by mayadebellis** [comment]
ssh working on ops!
